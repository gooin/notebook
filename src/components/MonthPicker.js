import React, {Component} from 'react';
import {Row, Col, Divider, Button} from 'antd';
import {add0WhenLess10, range} from '../utility';
import PropTypes from 'prop-types';


class MonthPicker extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isOpen: false,
            selectedYear: this.props.year,
            selectedMonth: this.props.month,
        }
    }

    componentDidMount() {
        document.addEventListener('click', this.handleClick, false);
    }

    componentWillUnmount() {
        document.removeEventListener('click', this.handleClick, false);
    }

    handleClick = (event) => {
        console.log("this", this);

        // 使用refs将dom节点挂载到this上
        if (this.node.contains(event.target)) {
            return;
        }
        this.setState({
            isOpen: false
        })
    }

    toggleDropdown = () => {
        this.setState({
            isOpen: !this.state.isOpen
        })
    }

    onChangeSelectedYear = (event, year) => {
        // event.preventDefault();
        this.setState({
            selectedYear: year
        })
    }

    onChangeSelectedMonth = (event, month) => {
        // event.preventDefault();
        this.setState({
            selectedMonth: month
        })
        this.toggleDropdown()
        this.props.onChange(this.state.selectedYear, month)
    }


    render() {
        const {year, month} = this.props
        // const year = this.state.selectedYear
        // const month = this.state.selectedMonth
        const {isOpen} = this.state
        const monthRange = range(12, 1);
        const yearRange = range(9, year - 4);
        return (
            <div>
                <div className="dropdown"
                    // 将此dom节点挂载到this上
                     ref={(ref) => {
                         this.node = ref
                     }}>
                    <h5>选择月份</h5>
                    <button className="btn btn-secondary dropdown-toggle month-selector" type="button"


                            onClick={this.toggleDropdown}
                    >
                        {`${year} 年 ${add0WhenLess10(month)} 月`}
                    </button>
                    {
                        isOpen &&
                        <div className="dropdown-menu"
                             style={{display: 'block', minWidth: '10rem', padding: '.5rem .5rem'}}>
                            {
                                monthRange.map((monthNumber, index) => (
                                    <Row key={index}>
                                        <Col span={9}>
                                            <Button
                                                ref={(ref) => {
                                                    // this.buttonNode[index] = ref
                                                }}
                                                type={this.state.selectedYear === yearRange[index] ? "primary" : 'dashed'}
                                                onClick={(event) => this.onChangeSelectedYear(event, yearRange[index])}
                                            >
                                                {yearRange[index]}
                                            </Button>
                                        </Col>
                                        <Divider type="vertical"/>
                                        <Col span={8}>
                                            <Button
                                                className={'month-button'}
                                                type={this.state.selectedMonth === monthNumber ? "primary" : 'dashed'}
                                                onClick={(event) => this.onChangeSelectedMonth(event, monthNumber)}
                                            >
                                                {add0WhenLess10(monthNumber)} 月
                                            </Button>
                                        </Col>
                                    </Row>
                                ))
                            }
                        </div>
                    }
                </div>
            </div>
        );
    }
}

MonthPicker.propTypes = {
    year: PropTypes.number.isRequired,
    month: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired
}


export default MonthPicker;