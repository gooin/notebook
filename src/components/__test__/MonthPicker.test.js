import React from 'react';
import ReactDOM from 'react-dom';
import {shallow,mount} from 'enzyme';
import MonthPicker from "../MonthPicker";
import toJson from 'enzyme-to-json';

// const jsdom = new JSDOM('<!doctype html><html><body></body></html>');
// const { window } = jsdom;
//
// function copyProps(src, target) {
//     Object.defineProperties(target, {
//         ...Object.getOwnPropertyDescriptors(src),
//         ...Object.getOwnPropertyDescriptors(target),
//     });
// }
//
// global.window = window;
// global.document = window.document;
// global.navigator = {
//     userAgent: 'node.js',
// };
// global.requestAnimationFrame = function (callback) {
//     return setTimeout(callback, 0);
// };
// global.cancelAnimationFrame = function (id) {
//     clearTimeout(id);
// };
// copyProps(window, global);


const props = {
    year: 2020,
    month: 11,
    onChange: jest.fn()
}


describe('test MonthPicker component', () => {


    let wrapper = shallow(<MonthPicker {...props}/>)


    it('should h5 rendered', () => {
        expect(wrapper.find('h5').length).toBe(1)
    })
    it('should button contains year and month', () => {
        expect(wrapper.find('button').text()).toContain('2020 年 11 月')
    })
    it('should default month render to 0', () => {
        expect(wrapper.find('Row').length).toBe(0)
    })
    it('click dropdown button, render year and month', () => {
        // 默认isOpen为false
        expect(wrapper.state('isOpen')).toBeFalsy()
        // 模拟点击，展开年月
        wrapper.find('button').simulate('click')
        // 展开后检查快照
        expect(toJson(wrapper)).toMatchSnapshot()
        // 点击后，isOpen为true
        expect(wrapper.state('isOpen')).toBeTruthy()
        // 检查是否渲染出来了12行
        expect(wrapper.find('Row').length).toBe(12)
        // 点击渲染出的第一个按钮，应该是2016
        wrapper.find('Button').first().simulate('click')
        expect(wrapper.state('selectedYear')).toEqual(2016)
        // 点击渲染出的最后一个按钮，月份应该是12
        wrapper.find('Button').last().simulate('click')
        expect(wrapper.state('selectedMonth')).toEqual(12)
        // 点击月份后会关闭面板,渲染的行数应该为0，面板状态应该为false
        expect(wrapper.find('Row').length).toBe(0)
        expect(wrapper.state('isOpen')).toBeFalsy()
        // 检查回调是否调用
        expect(props.onChange).toHaveBeenCalled()

    })

    it('click dropdown button, click document to close dropdown', () => {
        Object.defineProperty(window, "matchMedia", {
            writable: true,
            value: jest.fn().mockImplementation(query => ({
                matches: false,
                media: query,
                onchange: null,
                addListener: jest.fn(), // Deprecated
                removeListener: jest.fn(), // Deprecated
                addEventListener: jest.fn(),
                removeEventListener: jest.fn(),
                dispatchEvent: jest.fn(),
            }))
        });


        const wrapper1 = mount(<MonthPicker {...props}/>)
        // // 模拟点击，展开年月
        wrapper1.find('button').simulate('click')
        // 点击后，isOpen为true
        expect(wrapper1.state('isOpen')).toBeTruthy()
        // 检查是否渲染出来了12行
        expect(wrapper1.find('Row').length).toBe(12)

        // let eventMap = {}
        // document.addEventListener = jest.fn((event, cb) => {
        //     eventMap[event] = cb
        // })
        //
        // eventMap.click({
        //     target:ReactDOM.findDOMNode(wrapper1.instance())
        // })
        // // 点击后，isOpen为true
        // expect(wrapper1.state('isOpen')).toBeTruthy()


    })

})