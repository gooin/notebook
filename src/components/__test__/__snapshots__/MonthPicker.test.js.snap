// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MonthPicker component click dropdown button, render year and month 1`] = `
<div>
  <div
    className="dropdown"
  >
    <h5>
      选择月份
    </h5>
    <button
      className="btn btn-secondary dropdown-toggle"
      onClick={[Function]}
      type="button"
    >
      2020 年 11 月
    </button>
    <div
      className="dropdown-menu"
      style={
        Object {
          "display": "block",
          "minWidth": "10rem",
          "padding": ".5rem .5rem",
        }
      }
    >
      <Row
        key="0"
      >
        <Col
          span={9}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            2016
          </Button>
        </Col>
        <Divider
          type="vertical"
        />
        <Col
          span={8}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            01
             月
          </Button>
        </Col>
      </Row>
      <Row
        key="1"
      >
        <Col
          span={9}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            2017
          </Button>
        </Col>
        <Divider
          type="vertical"
        />
        <Col
          span={8}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            02
             月
          </Button>
        </Col>
      </Row>
      <Row
        key="2"
      >
        <Col
          span={9}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            2018
          </Button>
        </Col>
        <Divider
          type="vertical"
        />
        <Col
          span={8}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            03
             月
          </Button>
        </Col>
      </Row>
      <Row
        key="3"
      >
        <Col
          span={9}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            2019
          </Button>
        </Col>
        <Divider
          type="vertical"
        />
        <Col
          span={8}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            04
             月
          </Button>
        </Col>
      </Row>
      <Row
        key="4"
      >
        <Col
          span={9}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="primary"
          >
            2020
          </Button>
        </Col>
        <Divider
          type="vertical"
        />
        <Col
          span={8}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            05
             月
          </Button>
        </Col>
      </Row>
      <Row
        key="5"
      >
        <Col
          span={9}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            2021
          </Button>
        </Col>
        <Divider
          type="vertical"
        />
        <Col
          span={8}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            06
             月
          </Button>
        </Col>
      </Row>
      <Row
        key="6"
      >
        <Col
          span={9}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            2022
          </Button>
        </Col>
        <Divider
          type="vertical"
        />
        <Col
          span={8}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            07
             月
          </Button>
        </Col>
      </Row>
      <Row
        key="7"
      >
        <Col
          span={9}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            2023
          </Button>
        </Col>
        <Divider
          type="vertical"
        />
        <Col
          span={8}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            08
             月
          </Button>
        </Col>
      </Row>
      <Row
        key="8"
      >
        <Col
          span={9}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            2024
          </Button>
        </Col>
        <Divider
          type="vertical"
        />
        <Col
          span={8}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            09
             月
          </Button>
        </Col>
      </Row>
      <Row
        key="9"
      >
        <Col
          span={9}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          />
        </Col>
        <Divider
          type="vertical"
        />
        <Col
          span={8}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            10
             月
          </Button>
        </Col>
      </Row>
      <Row
        key="10"
      >
        <Col
          span={9}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          />
        </Col>
        <Divider
          type="vertical"
        />
        <Col
          span={8}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="primary"
          >
            11
             月
          </Button>
        </Col>
      </Row>
      <Row
        key="11"
      >
        <Col
          span={9}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          />
        </Col>
        <Divider
          type="vertical"
        />
        <Col
          span={8}
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="dashed"
          >
            12
             月
          </Button>
        </Col>
      </Row>
    </div>
  </div>
</div>
`;
